casper/src/main/scala/coop/rchain/casper/util/rholang/Checkpoint.scala

```diff
@@ -2,29 +2,67 @@ package coop.rchain.casper.util.rholang
 
 import com.google.protobuf.ByteString
 
-import coop.rchain.crypto.codec.Base16
+import coop.rchain.catscontrib.TaskContrib._
+import coop.rchain.models.Par
+import coop.rchain.rholang.interpreter.Runtime
+import coop.rchain.rholang.interpreter.storage.StoragePrinter
+import coop.rchain.rspace.Blake2b256Hash
 
-import java.nio.file.Path
+import monix.execution.Scheduler
 
-class Checkpoint(val hash: ByteString, val location: Path, val size: Long) {
-  val dbLocation: Path = location.resolve(Base16.encode(hash.toByteArray))
+import scala.concurrent.SyncVar
+import scala.util.{Failure, Success, Try}
 
-  def toTuplespace(name: String): Tuplespace = {
-    val tsLocation = location.resolve(name)
+//runtime is a SyncVar for thread-safety, as all checkpoints share the same "hot store"
+class Checkpoint private (val hash: ByteString, runtime: SyncVar[Runtime]) {
 
-    tsLocation.toFile.mkdir()
-    Tuplespace.copyDB(dbLocation, tsLocation)
+  def updated(terms: List[Par])(implicit scheduler: Scheduler): Either[Throwable, Checkpoint] = {
+    val hot   = getHot()
+    val error = eval(terms, hot)
+    val newHash = error.fold[Either[Throwable, ByteString]](
+      Right(ByteString.copyFrom(hot.space.getCheckpoint().bytes.toArray)))(Left(_))
+    runtime.put(hot)
 
-    new Tuplespace(name, location, size)
+    newHash.map(new Checkpoint(_, runtime))
   }
 
-  def toTuplespace: Tuplespace = {
-    val name       = Tuplespace.randomName
-    val tsLocation = location.resolve(name)
+  def storageRepr: String = {
+    val hot    = getHot()
+    val result = StoragePrinter.prettyPrint(hot.space.store)
+    runtime.put(hot)
+    result
+  }
+
+  private def getHot(): Runtime = {
+    val hot       = runtime.take()
```

Comment:
This "hot" is the state right? Equivalent to what you get in a working directory with a git checkout.

Reply:
Is `Runtime` _hot_ or rather `rspace`? I have thought that hot vs cold was the terminology for storage.

Reply:
Yeah, maybe the nomenclature is a little confusing. Maybe `active` would be better? Open to suggestions on this front :)

Reply:
@KentShikama yes, `hot` is like the working directory in git. And the `getHot` function returns a pointer to that working directory and resets it to the checkpoint's state as a side-effect.

Reply:
I got bit confused and I started thinking that `Runtime` can be `hot` and `cold` like the `rspace` which made me worry because I had not heard about that. `active` sounds better.